### PR TITLE
Add the utf8 library

### PR DESCRIPTION
**Pull request by plt-abigail**
In start contrast to my other two PRs that were at most 5 lines of changes, here's the _entire_ Lua 5.3 `utf8` library.

### CODE REVIEW

**Comment by plt-abigail** [comment]
One of those commits is in the wrong branch..

**Comment by plt-abigail** [comment]
Fixed.

**Comment by plt-abigail** [comment]
Damn it, I can't turn a PR into a draft, but please don't merge this until we write tests for it :)

File: src/main/java/org/squiddev/cobalt/lib/Utf8Lib.java [hunk 1]
```diff
@@ -0,0 +1,187 @@
...
+			if (n < 0) {
+				n = 0;
+			} else if (n < s.length) {
+				n++;
+				while (isCont(s, n)) n++;
+			}
+			if (n >= s.length) {
+				return ValueFactory.varargsOf();
+			} else {
+				long c = decodeUtf8(s, n, new int[0]);
```
[the remark this answers is not in the document]
Comment (plt-abigail) [reply src/main/java/org/squiddev/cobalt/lib/Utf8Lib.java]: Good point, but it never crashed on me. I'll change anyway

File: src/main/java/org/squiddev/cobalt/Buffer.java [hunk 2]
```diff
@@ -156,6 +156,19 @@ public final Buffer append(byte b) {
...
 
+	/**
+	 * Append an array of bytes to the buffer.
+	 *
+	 * @param b The bytes to append
+	 * @return {@code this} to allow call chaining
+	 */
+	public final Buffer append(byte[] b) {
+		makeRoom(0, b.length);
+		for(int i = 0; i < b.length; i++)
```
[the remark this answers is not in the document]
Comment (plt-abigail) [reply src/main/java/org/squiddev/cobalt/Buffer.java]: :+1: I'll do it as soon as IDEA opens, i.e., in 3 months

File: src/main/java/org/squiddev/cobalt/Buffer.java [hunk 3]
```diff
@@ -156,6 +156,20 @@ public final Buffer append(byte b) {
...
+	/**
+	 * Append an array of bytes to the buffer.
+	 *
+	 * @param b The bytes to append
+	 * @return {@code this} to allow call chaining
+	 */
+	public final Buffer append(byte[] b) {
+		makeRoom(0, b.length);
+		System.arraycopy(b, 0, bytes, offset, b.length);
+		offset += b.length;
```
Comment (SquidDev) [inline src/main/java/org/squiddev/cobalt/Buffer.java]: I don't think offset should grow, should it?

Comment (plt-abigail) [reply src/main/java/org/squiddev/cobalt/Buffer.java]: Sorry, I thought offset was a pointer to the "write head". Fixed

File: src/main/java/org/squiddev/cobalt/lib/Utf8Lib.java [hunk 4]
```diff
@@ -149,26 +154,21 @@ private static int posRelative(int pos, int len) {
...
-			s = bytes;
-		}
-
 		@Override
 		public Varargs invoke(LuaState state, Varargs args) throws LuaError, UnwindThrowable {
-			// Arg 1: invariant state (nil)
+			// Arg 1: invariant state (the string)
 			// Arg 2: byte offset + 1
 			// Returns: byte offset + 1, code point
+			byte[] s = args.arg(1).checkLuaString().bytes;
```
Comment (SquidDev) [inline src/main/java/org/squiddev/cobalt/lib/Utf8Lib.java]: Oh goodness, I should have noticed this earlier sorry! This isn't entirely correct: `LuaString`s are a range over that, so `s.luaByte(i) != s.bytes[i]` (and `s.length != s.bytes.length`).

The sanest thing to do here might be to replace `byte[]` with `LuaString` everywhere, and then use `.substring` for `codepoints` - we can optimise it later.

Sorry, should have noticed this all before 🤦‍♂.

Comment (plt-abigail) [reply src/main/java/org/squiddev/cobalt/lib/Utf8Lib.java]: Alright, I'm on it :+1:

**Pull request merged**